src/Compilers/Core/Portable/CodeGen/PrivateImplementationDetails.cs

```diff
@@ -292,8 +311,77 @@ internal Cci.IFieldReference CreateDataField(ImmutableArray<byte> data, ushort a
                     _ => hex
                 };
 
-                return new MappedField(name, this, type, key.Data);
-            });
+                return new MappedField(name, @this, type, data);
+            },
+            this);
+        }
+
+        /// <summary>
+        /// Gets the field of <see cref="DataSectionStringType"/> or creates one
+        /// if the type does not exist yet for the given <paramref name="text"/>.
+        /// </summary>
+        internal Cci.IFieldReference GetOrCreateFieldForStringValue(
+            string text,
+            ImmutableArray<byte> data,
```

Comment:
>ImmutableArray<byte> data, [](http://example.com/codeflow?start=12&length=26)

We still keep passing redundant data. I understand that we want to avoid creation of private implementation detail type when a string cannot be encoded. This can be achieved by making this a static method and passing CommonPEModuleBuilder in. The method can attempt the encoding, request the instance of the type from the builder and complete the work.